gitpod-io/gitpod
# [supervisor] tasks support

#### What it does

- Supervisor provides an API to get the current state of Gitpod tasks. Each task:
  - has a unique ID which is stable between workspace restarts based on order in GITPOD_TASKS env variable;
    - Theia now uses such stable ID to reliably match frontend terminals to remote instead of guessing between workspace restarts. Default user terminals follow the default Theia logic without any magic reordering.
  - can be either in opening, running or closed state:
    - If a task must be executed in the current workspace then it is initialised in opening state, otherwise in closed;
      - Theia creates frontend terminals for opening remote terminals and ignores initially closed.
    - As soon as the supervisor manages to create a terminal a task moves to the running state or closed state if it fails;
      - Theia attaches frontend terminals to remote terminals using `supervisor terminal attach` if it is not attached yet (in case of page reload for instance);
    - When an underlying terminal process exits a task moves to the closed state.
      - Theia closes a frontend terminal if a remote terminal is closed and another way around, i.e. if a user closes a frontend terminal then the supervisor is notified to close a corresponding remote terminal as well.
  - comes with presentation hints which is used by Theia to layout frontend terminals.

#### How to test

During testing use a workspace with multiple tasks having presentation hints, for instance:
- regular - http://ak-supervisor-tasks.staging.gitpod-dev.com/#https://github.com/akosyakov/theia-training/tree/solution-1
- prebuild - http://ak-supervisor-tasks.staging.gitpod-dev.com/#prebuild/https://github.com/akosyakov/theia-training/tree/solution-1

Tests:
- workspace creation: 
  - start - check that init phase is running, terminal output and history should be proper
  - restart - check that init phase is skipped, terminal output and history should be proper
  - prebuild - logs should contain workspace events
  - prebuilt - prebuild logs are displayed, how much time was saved should be reported
- terminal layout restoration:
  - create a new workspace with tasks and open some additional terminals
  - check that layout is proper and relayout afterward
  - check that layout is kept after reload and restart
  - check that after reload all terminals are matched to the same backend processes as before reload
  - check that after restart task terminals are matched to proper remote tasks and remaining terminals are connected to new processes
- terminal lifecycle
  - close a terminal with task and use `/theia/supervisor terminal list` to check that an underlying task is closed as well
  - create a new terminal, use `/theia/supervisor terminal attach -ir ` to attach to existing task and `exit` it, check that other terminals get closed as well
- reconnection
  - run some command in the terminal which constantly produces the output: `watch -n 0.01 ps -auxf`
  - emulate network issues, i.e. use Network Link Conditioner on Mac to enable 100% data loss on system level, don't use Chrome for it, it cannot throttle web sockets
  - disable emulation after seeing the yellow statusbar, observe whether terminals are reconnected, operable and output is good
- test Code as well:
  - grant devops role to your user, i.e. run a query like: update d_b_user set rolesOrPermissions = '["admin","devops"]' where name = 'akosyakov';
  - start workspace http://ak-supervisor-tasks.staging.gitpod-dev.com/#https://github.com/akosyakov/go-gin-app
  - Code generally is not capable to preserve terminal on the page reload, but tasks terminals should be preserved
